node/actors/bft/src/chonky_bft/proposal.rs

```diff
@@ -0,0 +1,233 @@
+        match self {
+            Error::Internal(err) => Error::Internal(err.with_wrap(f)),
+            err => err,
+        }
+    }
+}
+
+impl StateMachine {
+    /// Processes a LeaderProposal message.
+    pub(crate) async fn on_proposal(
+        &mut self,
+        ctx: &ctx::Ctx,
+        signed_message: validator::Signed<validator::LeaderProposal>,
+    ) -> Result<(), Error> {
+        // ----------- Checking origin of the message --------------
+
+        // Unwrap message.
+        let message = &signed_message.msg;
+        let author = &signed_message.key;
+        let view = message.view().number;
+
+        // Check that the message is for the current view or a future view. We only allow proposals for
+        // the current view if we have not voted or timed out yet.
+        if view < self.view_number
+            || (view == self.view_number && self.phase != validator::Phase::Prepare)
+        {
+            return Err(Error::Old {
+                current_view: self.view_number,
+                current_phase: self.phase,
+            });
+        }
+
+        // Check that it comes from the correct leader.
+        let leader = self.config.genesis().view_leader(view);
+        if author != &leader {
+            return Err(Error::InvalidLeader {
+                correct_leader: leader,
+                received_leader: author.clone(),
+            });
+        }
+
+        // ----------- Checking the message --------------
+
+        signed_message.verify().map_err(Error::InvalidSignature)?;
+
+        message
+            .verify(self.config.genesis())
+            .map_err(Error::InvalidMessage)?;
+
+        let (implied_block_number, implied_block_hash) = message
+            .justification
+            .get_implied_block(self.config.genesis());
+
+        // Replica MUSTN'T vote for blocks which have been already pruned for storage.
+        // (because it won't be able to persist and broadcast them once finalized).
+        // TODO(gprusak): it should never happen, we should add safety checks to prevent
+        // pruning blocks not known to be finalized.
+        if implied_block_number < self.config.block_store.queued().first {
+            return Err(Error::ProposalAlreadyPruned);
+        }
+
+        let block_hash = match implied_block_hash {
+            // This is a reproposal. We let the leader repropose blocks without sending
+            // them in the proposal (it sends only the number + hash). That allows a
+            // leader to repropose a block without having it stored.
+            // It is an optimization that allows us to not wait for a leader that has
+            // the previous proposal stored (which can take 4f views), and to somewhat
+            // speed up reproposals by skipping block broadcast.
+            // This only saves time because we have a gossip network running in parallel,
+            // and any time a replica is able to create a finalized block (by possessing
```

Comment:
This comment is misleading. Reproposals are critical for liveness, they are not an optimizations. We repropose without payload, because we don't gossip non-finalized payloads (and that part is indeed an optimization).

Reply:
Ok, I'll reword it to be more clear.